opencog/query/DefaultPatternMatchCB.cc

```diff
@@ -389,33 +391,34 @@ bool DefaultPatternMatchCB::is_self_ground(const Handle& ptrn,
                                            const Handle& grnd,
                                            const HandleMap& term_gnds,
                                            const OrderedHandleSet& varset,
-                                           int quote_level)
+                                           Quotation quotation)
 {
 	Type ptype = ptrn->getType();
 
 	// Unwrap quotations, so that they can be compared properly.
-	if (ptype == QUOTE_LINK or ptype == UNQUOTE_LINK)
+	if (Quotation::is_quotation_type(ptype))
 	{
 		// Wow, if we are here, and patern==grnd, this must be
-		// a self-grounding, as I don't beleive there is any other
+		// a self-grounding, as I don't believe there is any other
 		// valid way to get to here.
-		if (ptype == QUOTE_LINK and 0 == quote_level and ptrn == grnd) return true;
-		if (ptype == UNQUOTE_LINK and 1 == quote_level and ptrn == grnd) return true;
-
-		if (ptype == QUOTE_LINK) quote_level++;
-		else quote_level--;
+		if (quotation.consumable(ptype) and ptrn == grnd) return true;
 
 		const Handle& qpat = ptrn->getOutgoingAtom(0);
-		return is_self_ground(qpat, grnd, term_gnds, varset, quote_level);
```

Comment:
This is confusing. Up above, the unwrapping of the quote link was done in the pattern term (line 49) -- while here, it is unwrapped a second time?  Or maybe the outgoing set is no longer the outgoing set of the atom, but is instead the outgoing set of the quote?  

That is, it seems that `ptrn->getOutgoingAtom(0)`  has the same value as `ptrn->getHandle()` whenever the ptrn is a QuoteLink. This seems like a kind-of schizo behavior in the code ...